internal/rfc7512/rfc7512.go

```diff
@@ -0,0 +1,182 @@
+	r, _ := regexp.Compile("^pkcs11:" + pClause + "(\\?" + qClause + ")?$")
+
+	if !r.MatchString(uri) {
+		return nil, fmt.Errorf("PKCS#11 URI is malformed: %s", uri)
+	}
+
+	var pk11uri PKCS11URI
+	var pAttrs []string
+	var qAttrs []string
+	var parts []string
+	var value string
+
+	// Separate the scheme name, path, and query attributes:
+	uri = strings.Split(uri, "pkcs11:")[1]
+	parts = strings.Split(uri, "?")
+	pAttrs = strings.Split(parts[0], ";")
+	if 1 < len(parts) {
+		qAttrs = strings.Split(parts[1], "&")
+	}
+
+	// Parse the path attributes:
+	for _, attr := range pAttrs {
+		parts = strings.Split(attr, "=")
+		parts[0] = strings.Trim(parts[0], " \n\t\r")
+		if len(parts) == 2 {
+			parts[1] = strings.Trim(parts[1], " \n\t\r")
+			value, _ = url.QueryUnescape(parts[1])
+		} else {
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Path Attribute: %s", parts[0])
+		}
+		switch parts[0] {
+		case "token":
+			pk11uri.Token = value
+		case "manufacturer":
+			pk11uri.Manuf = value
+		case "serial":
+			pk11uri.Serial = value
+		case "model":
+			pk11uri.Model = value
+		case "library-manufacturer":
+			pk11uri.LibManuf = value
+		case "library-description":
+			pk11uri.LibDesc = value
+		case "library-version":
+			// TODO 1*DIGIT [ "." 1*DIGIT ]
+			pk11uri.LibVer = value
+		case "object":
+			pk11uri.Object = []byte(value)
+		case "type":
+			// TODO public, private, cert, secret-key, data
+			pk11uri.Type = value
+		case "id":
+			pk11uri.ID = []byte(value)
+		case "slot-manufacturer":
+			pk11uri.SlotManuf = value
+		case "slot-description":
+			pk11uri.SlotDesc = value
+		case "slot-id":
+			// TODO the bit size is not clarified.
+			id, _ := strconv.ParseUint(value, 10, 32)
+			pk11uri.SlotID = uint(id)
+		default:
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Path Attribute: %s", parts[0])
+		}
+	}
+
+	// Parse the query attributes:
+	for _, attr := range qAttrs {
+		parts = strings.Split(attr, "=")
+		parts[0] = strings.Trim(parts[0], " \n\t\r")
+		if len(parts) == 2 {
+			parts[1] = strings.Trim(parts[1], " \n\t\r")
+			value, _ = url.QueryUnescape(parts[1])
+		} else {
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Query Attribute: %s", parts[0])
+		}
+		switch parts[0] {
+		case "pin-source":
+			pk11uri.PinSource = value
+		case "pin-value":
+			pk11uri.PinValue = value
+		case "module-name":
+			pk11uri.ModuleName = value
+		case "module-path":
+			pk11uri.ModulePath = value
+		case "max-sessions":
+			maxSessions, _ := strconv.ParseUint(value, 10, 32)
+			pk11uri.MaxSessions = int(maxSessions)
+		default:
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Query Attribute: %s", parts[0])
+		}
+	}
+
+	return &pk11uri, nil
+}
+
+// LoadPKCS11Key attempts to load a Signer given a PKCS11URI object identifier.
+// An error is returned if the crypto11 module cannot find the module, token,
+// or the specified object.
+func LoadPKCS11Key(pk11uri *PKCS11URI) (crypto.Signer, error) {
```

Comment:
> We're not doing any of that. We're simply calling a function in crypto11 to __load a key from a PKCS11 module__ and give it to us as a crypto.Signer object. I don't think the function name could be more declarative.

Please expand on the bolded section. "module, token, or the specified object" -- I don't know what any of these words mean.